From 2f55b8ec68e4ec0b2adf5ba6733e76529295b2b0 Mon Sep 17 00:00:00 2001
From: Ruchi Kandoi <kandoiruchi@google.com>
Date: Wed, 17 Oct 2018 16:14:08 -0700
Subject: [PATCH 4/4] Prevent OOB error in nfc_ncif_proc_get_routing()

Test: Tag reading; Card Emulation
Bug: 117554809
Change-Id: Ib49af2eadf870f030a6cddeec390dc498bd5078c
(cherry picked from commit ded496ea745656018dda505c23726b4304180c38)
---
 src/nfc/nfc/nfc_ncif.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/nfc/nfc/nfc_ncif.c b/src/nfc/nfc/nfc_ncif.c
index 6f0f429..45aac6a 100644
--- a/src/nfc/nfc/nfc_ncif.c
+++ b/src/nfc/nfc/nfc_ncif.c
@@ -24,6 +24,7 @@
  *  (callback). On the transmit side, it manages the command transmission.
  *
  ******************************************************************************/
+#include <log/log.h>
 #include <stdlib.h>
 #include <string.h>
 #include "nfc_target.h"
@@ -1191,8 +1192,14 @@ void nfc_ncif_proc_get_routing (UINT8 *p, UINT8 len)
             {
                 tl                  = *(p+1);
                 tl                 += NFC_TL_SIZE;
-                STREAM_TO_ARRAY (pn, p, tl);
                 evt_data.tlv_size  += tl;
+                if (evt_data.tlv_size > NFC_MAX_EE_TLV_SIZE)
+                {
+                    android_errorWriteLog(0x534e4554, "117554809");
+                    NFC_TRACE_ERROR1 ("%s: Invalid data format", __func__);
+                    return;
+                }
+                STREAM_TO_ARRAY (pn, p, tl);
                 pn                 += tl;
             }
             (*nfc_cb.p_resp_cback) (NFC_GET_ROUTING_REVT, (tNFC_RESPONSE *) &evt_data);
-- 
2.7.4

